planner/core/logical_plan_builder.go

```diff
@@ -856,6 +860,152 @@ func (b *PlanBuilder) buildJoin(ctx context.Context, joinNode *ast.Join) (Logica
+
+	// Merge sub-plan's fullNames into this join plan, similar to the fullSchema logic above.
+	leftJoinPlan.fullNames = make([]*types.FieldName, 0, len(lFullNames)+len(rFullNames))
+	for _, lName := range lFullNames {
+		name := *lName
+		leftJoinPlan.fullNames = append(leftJoinPlan.fullNames, &name)
+	}
+	for _, rName := range rFullNames {
+		name := *rName
+		leftJoinPlan.fullNames = append(leftJoinPlan.fullNames, &name)
+	}
+
+	// Clear NotNull flag for the inner side schema if it's an outer join.
+	if joinNode.Tp == ast.LeftJoin || joinNode.Tp == ast.RightJoin {
+		resetNotNullFlag(leftJoinPlan.fullSchema, 0, leftJoinPlan.fullSchema.Len())
+	}
+
+	if joinNode.On != nil {
+		b.curClause = onClause
+		onExpr, newPlan, err := b.rewrite(ctx, joinNode.On.Expr, leftJoinPlan, nil, false)
+		if err != nil {
+			return nil, err
+		}
+		if newPlan != leftJoinPlan {
+			return nil, errors.New("ON condition doesn't support subqueries yet")
+		}
+		onCondition := expression.SplitCNFItems(onExpr)
+		leftJoinPlan.AttachOnConds(onCondition)
+	}
+
+	// build the part of anti join.
+	lChild2, err := b.buildResultSetNode(ctx, joinNode.Left, false)
+	if err != nil {
+		return nil, err
+	}
+	rChild2, err := b.buildResultSetNode(ctx, joinNode.Right, false)
+	if err != nil {
+		return nil, err
+	}
+	handleMap1 = b.handleHelper.popMap()
+	handleMap2 = b.handleHelper.popMap()
+	b.handleHelper.mergeAndPush(handleMap1, handleMap2)
+
+	antiJoinPlan := LogicalJoin{StraightJoin: joinNode.StraightJoin || b.inStraightJoin}.Init(b.ctx, b.getSelectOffset())
+	antiJoinPlan.SetChildren(rChild2, lChild2)
+	antiJoinPlan.SetSchema(expression.MergeSchema(rChild2.Schema(), lChild2.Schema()))
+	antiJoinPlan.names = make([]*types.FieldName, rChild2.Schema().Len()+lChild2.Schema().Len())
+	copy(antiJoinPlan.names, rChild2.OutputNames())
+	copy(antiJoinPlan.names[rChild2.Schema().Len():], lChild2.OutputNames())
+	antiJoinPlan.JoinType = AntiSemiJoin
+	if joinNode.On != nil {
+		b.curClause = onClause
+		onExpr, newPlan, err := b.rewrite(ctx, joinNode.On.Expr, antiJoinPlan, nil, false)
+		if err != nil {
+			return nil, err
+		}
+		if newPlan != antiJoinPlan {
+			return nil, errors.New("ON condition doesn't support subqueries yet")
+		}
+		onCondition := expression.SplitCNFItems(onExpr)
+		antiJoinPlan.AttachOnConds(onCondition)
+	}
+	antiJoinPlan.SetSchema(rChild2.Schema())
+	antiJoinPlan.SetOutputNames(rChild2.OutputNames())
+
+	// Supplement the null parts.
+	projExprs := make([]expression.Expression, lChild2.Schema().Len()+rChild2.Schema().Len())
+	copy(projExprs[lChild2.Schema().Len():], expression.Column2Exprs(rChild2.Schema().Columns))
+	for i := 0; i < lChild2.Schema().Len(); i++ {
+		projExprs[i] = expression.NewNull()
+	}
+	proj := LogicalProjection{
+		Exprs: projExprs,
+	}.Init(b.ctx, b.getSelectOffset())
+	proj.SetChildren(antiJoinPlan)
+	proj.SetSchema(expression.MergeSchema(lChild2.Schema(), rChild2.Schema()))
```

Comment:
Looks like we should also reset the NotNULL flag for the `lChild2` schema in the projection schema here.

Reply:
I'm going to reset the flags for all columns.